### ISSUE TITLE
Client: "Step must only increase in log calls" solution is to not set "step"

### ISSUE THREAD

**Issue by vib2810** [wandb/wandb#2258]
**Description**
- When you stop the wandb script, the step count is always one number higher than the epoch count (I log values at every epoch).![image](https://user-images.githubusercontent.com/35529327/120993023-c9d7f400-c7a0-11eb-9465-1c89b3375b31.png) (Here step=3 but the last log was at epoch=step=2)
- This is because of wandb.watch() which logs values at step=3 even though no value is logged at that step. (Which can be clearly seen in the image below (Gradients logged till step=3, But the loss was logged only till step=2 ![image](https://user-images.githubusercontent.com/35529327/120976428-f6374480-c78f-11eb-920d-0edec0b49387.png) ![image](https://user-images.githubusercontent.com/35529327/120973872-1e717400-c78d-11eb-91a9-f43b9bb8c26f.png)
- This causes huge problems when resuming the script again as we cant log values at step=3 after resuming (the error step 3<4) ![image](https://user-images.githubusercontent.com/35529327/120974731-1b2ab800-c78e-11eb-8ce2-7c80847402b9.png) ![image](https://user-images.githubusercontent.com/35529327/120975546-faaf2d80-c78e-11eb-8227-ff2ee82c5957.png)
(Error while resuming. Notice step increased to 4 from 3, as wandb.watch() logged gradients for step=4 now)

**Wandb features**
```
 if(args['resume']==True):
        wandb.init(project="NAME", entity='XXX', id=args['resume_run_id'], resume='must') 
    else:
        wandb.init(project="NAME", entity='XXX', config=config, name=args['run_name'])  
```
Before the train loop
`    wandb.watch(model, criterion, log="all", log_freq=1)`

**How to reproduce**
Simply run a model with wandb.watch(), and log with wandb.log({'epoch'=epoch, ......}, step=epoch), and terminate after the script with ctrl+C after an epoch. You will find a mismatch between the _step logged and epoch logged. 

**Environment**
- OS: Ubuntu 18.04
- Library: Pytorch
- Python Version: Python 3.8.5

**Possible Fix**
wandb.watch() will log values at a particular step only after wandb.log() is called at that step

**Comment by vanpelt** [comment]
Hey @vib2810 we suggest not using the `step` argument in this case.  You can log a custom step as one of the metrics in your wandb.log calls.  A future release may support steps going back in time, but currently does not.

**Comment by vib2810** [comment]
Hey @vanpelt thanks for the response.
So what you are suggesting is
`wanbd.log({epoch=.., others values to log}`
Since no step argument is given, it will log at the next step. Although this will create an offset between epoch and step, but ya is a possible temporary fix. 
Looking forward to allowing logging at past steps as well
Thankyou
